### ISSUE TITLE
Hints based autodiscover CheckConfig() fails due to race condition and/or missing error check

### ISSUE THREAD

**Issue by manzsolutions-lpr** [elastic/beats#23883]
```
ERROR	fileset/factory.go:103	Error creating input: Can only start an input when all related states are finished […]
ERROR	[autodiscover]	autodiscover/autodiscover.go:209	Auto discover config check failed for config […]
```

This only happens when Filebeat is already running when the annotated pod is started. If Filebeat is restarted this error does not appear and the nginx module works as expected.

In v7.10.2 the flow for autodiscovery seems to be as follows using bits of information from the above log:

autodiscover.go:207 --> factory.CheckConfig() -->
filebeat/fileset/factory.go:119 --> f.Create()

The important part here is that unlike CheckConfig() in filebeat/input/runnerfactory.go:60 the fileset factory CheckConfig() does not ignore common.ErrInputNotFinished causing the CheckConfig() to fail.

I see two issues here:

1. A probable race condition since the container apparently has an open input before the hints-based configuration is applied.
2. CheckConfig() does not ignore an error irrelevant to the check

Probably sufficient fix: Copy & paste filebeat/input/runnerfactory.go:CheckConfig():

```go
	if _, ok := err.(*common.ErrInputNotFinished); ok {
		// error is related to state, and hence config can be considered valid
		return nil
	}
```

- Version: 7.10.2
- Operating System: ECK
- Steps to Reproduce: 
1. Deploy Filebeat via ECK with hints based autodiscover activated

```yaml
spec:
  type: filebeat
  version: 7.10.2
  elasticsearchRef:
    name: elasticsearch
  kibanaRef:
    name: kibana
  config:
    filebeat:
      autodiscover:
        providers:
        - type: kubernetes
          node: ${NODE_NAME}
          hints:
            enabled: true
            default_config:
              type: container
              paths:
              - /var/log/containers/*${data.kubernetes.container.id}.log
    processors:
    - add_cloud_metadata: {}
    - add_host_metadata: {}
  daemonSet:
    podTemplate:
      spec:
        serviceAccountName: elastic-beat-filebeat
        automountServiceAccountToken: true
        dnsPolicy: ClusterFirstWithHostNet
        hostNetwork: true
        containers:
        - name: filebeat
          securityContext:
            runAsUser: 0
          volumeMounts:
          - name: var-log-containers
            mountPath: /var/log/containers
            readOnly: true
          - name: var-log-pods
            mountPath: /var/log/pods
            readOnly: true
          - name: var-lib-docker-containers
            mountPath: /var/lib/docker/containers
            readOnly: true
          env:
          - name: NODE_NAME
            valueFrom:
              fieldRef:
                fieldPath: spec.nodeName
        tolerations:
        - effect: "NoSchedule"
          operator: "Exists"
        volumes:
        - name: var-log-containers
          hostPath:
            path: /var/log/containers
        - name: var-log-pods
          hostPath:
            path: /var/log/pods
        - name: var-lib-docker-containers
          hostPath:
            path: /var/lib/docker/containers
```

2. Deploy ingress-nginx annotated with:
```yaml
  podAnnotations:
    co.elastic.logs/fileset.stdout: ingress_controller
    co.elastic.logs/module: nginx
```
3. Check filebeat log
```
2021-02-05T15:10:22.881Z	INFO	log/input.go:157	Configured paths: [/var/log/containers/*9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log]
2021-02-05T15:10:22.882Z	INFO	log/input.go:157	Configured paths: [/var/log/containers/*9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log]
2021-02-05T15:10:22.884Z	INFO	log/input.go:157	Configured paths: [/var/log/containers/*9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log]
2021-02-05T15:10:22.884Z	INFO	log/input.go:157	Configured paths: [/var/log/containers/*9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log]
2021-02-05T15:10:22.884Z	INFO	eslegclient/connection.go:99	elasticsearch url: https://elasticsearch-es-http.logs.svc:9200
2021-02-05T15:10:22.939Z	INFO	[esclientleg]	eslegclient/connection.go:314	Attempting to connect to Elasticsearch version 7.10.2
2021-02-05T15:10:22.944Z	INFO	log/harvester.go:302	Harvester started for file: /var/log/containers/nginx-76f9897ff5-nqwnf-nginx-9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log
2021-02-05T15:10:22.944Z	INFO	log/harvester.go:302	Harvester started for file: /var/log/containers/nginx-76f9897ff5-nqwnf-nginx-9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log
2021-02-05T15:10:23.959Z	INFO	input/input.go:136	input ticker stopped
2021-02-05T15:10:23.959Z	INFO	log/harvester.go:329	Reader was closed: /var/log/containers/nginx-76f9897ff5-nqwnf-nginx-9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log. Closing.
2021-02-05T15:10:23.959Z	INFO	input/input.go:136	input ticker stopped
2021-02-05T15:10:23.959Z	INFO	log/harvester.go:329	Reader was closed: /var/log/containers/nginx-76f9897ff5-nqwnf-nginx-9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log. Closing.
2021-02-05T15:10:23.960Z	ERROR	fileset/factory.go:103	Error creating input: Can only start an input when all related states are finished: {Id: d1307a69d7b00e1d-native::661536-64528, Finished: false, Fileinfo: &{nginx-76f9897ff5-nqwnf-nginx-9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log 0 416 {849094062 63748134619 0x64d0ce0} {64528 661536 1 33184 0 0 0 0 0 4096 0 {1612537819 849094062} {1612537819 849094062} {1612537819 849094062} [0 0 0]}}, Source: /var/log/containers/nginx-76f9897ff5-nqwnf-nginx-9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log, Offset: 0, Timestamp: 2021-02-05 15:10:23.696245126 +0000 UTC m=+612077.149984105, TTL: -1ns, Type: container, Meta: map[stream:stdout], FileStateOS: 661536-64528}
2021-02-05T15:10:23.960Z	ERROR	[autodiscover]	autodiscover/autodiscover.go:209	Auto discover config check failed for config '{
  "access": {
    "enabled": false,
    "input": {
      "paths": [
        "/var/log/containers/*9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log"
      ],
      "stream": "all",
      "type": "container"
    }
  },
  "error": {
    "enabled": false,
    "input": {
      "paths": [
        "/var/log/containers/*9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log"
      ],
      "stream": "all",
      "type": "container"
    }
  },
  "ingress_controller": {
    "enabled": true,
    "input": {
      "paths": [
        "/var/log/containers/*9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log"
      ],
      "stream": "stdout",
      "type": "container"
    }
  },
  "module": "nginx"
}', won't start runner: Can only start an input when all related states are finished: {Id: d1307a69d7b00e1d-native::661536-64528, Finished: false, Fileinfo: &{nginx-76f9897ff5-nqwnf-nginx-9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log 0 416 {849094062 63748134619 0x64d0ce0} {64528 661536 1 33184 0 0 0 0 0 4096 0 {1612537819 849094062} {1612537819 849094062} {1612537819 849094062} [0 0 0]}}, Source: /var/log/containers/nginx-76f9897ff5-nqwnf-nginx-9b3419089720ac95cd500c3ca2b4956c9cedecf06dac8c6cb9777e1e025a793c.log, Offset: 0, Timestamp: 2021-02-05 15:10:23.696245126 +0000 UTC m=+612077.149984105, TTL: -1ns, Type: container, Meta: map[stream:stdout], FileStateOS: 661536-64528}
```

**Comment by ChrsMark** [comment]
Hi @manzsolutions-lpr! Is this still valid? Did you try with a newer version of Filebeat to see if the issue still exists?

**Comment by manzsolutions-lpr** [comment]
@ChrsMark Thank you for motivating me to actually check again after such a long time. It actually appears to be solved at least in 7.16.3:

```
16:21:28.555Z	ERROR	[autodiscover]	cfgfile/list.go:99	Error creating runner from config: failed to create input: Can only start an input when all related states are finished: {Id: ea745ab688be85a9-native::1052673-64528, Finished: false, Fileinfo: […], Source: /var/log/containers/nginx_controller-<pod hash>.log, Offset: 721, Timestamp: 2022-02-01 16:21:09.958194363 +0000 UTC m=+6403.498933850, TTL: -1ns, Type: container, Meta: map[stream:stdout], FileStateOS: 1052673-64528}
16:21:38.655Z	INFO	[input]	log/input.go:171	Configured paths: [/var/log/containers/*<pod hash>.log]	{"input_id": "10f84fc1-cecb-486c-b3f6-db330b904c98"}
16:21:38.655Z	INFO	[esclientleg]	eslegclient/connection.go:102	elasticsearch url: https://elasticsearch-es-http.log-namespace.svc:9200
16:21:38.662Z	INFO	[esclientleg]	eslegclient/connection.go:282	Attempting to connect to Elasticsearch version 7.16.3
16:21:38.668Z	INFO	[input.harvester]	log/harvester.go:309	Harvester started for file.	{"input_id": "10f84fc1-cecb-486c-b3f6-db330b904c98", "source": "/var/log/containers/nginx_controller-<pod hash>.log", "state_id": "ea745ab688be85a9-native::1052673-64528", "finished": false, "os_id": "1052673-64528", "old_source": "/var/log/containers/nginx_controller-<pod hash>.log", "old_finished": true, "old_os_id": "1052673-64528", "harvester_id": "74daab46-a8ad-43b9-93a1-9cdce06041d5"}
```

I’ll keep my eyes open but I think it’s safe to close this issue.
